pkg/apis/resource/types.go

```diff
@@ -0,0 +1,449 @@
+	// PodClaimName matches the PodClaim.Name field.
+	PodClaimName string
+
+	// UnsuitableNodes lists nodes that the claim cannot be allocated for.
+	// Nodes listed here will be ignored by the scheduler when selecting a
+	// node for a Pod. All other nodes are potential candidates, either
+	// because no information is available yet or because allocation might
+	// succeed.
+	//
+	// A change of the PodSchedulingSpec.PotentialNodes field and/or a failed
+	// allocation attempt trigger an update of this field: the driver
+	// then checks all nodes listed in PotentialNodes and UnsuitableNodes
+	// and updates UnsuitableNodes.
+	//
+	// It must include the prior UnsuitableNodes in this check because the
+	// scheduler will not list those again in PotentialNodes but they might
+	// still be unsuitable.
+	//
+	// This can change, so the driver also must refresh this information
+	// periodically and/or after changing resource allocation for some
+	// other ResourceClaim until a node gets selected by the scheduler.
+	//
+	// The size of this field is limited to 128, the same as for
+	// PodSchedulingSpec.PotentialNodes.
+	UnsuitableNodes []string
+}
+
+// PodSchedulingNodeListMaxSize defines the maximum number of entries in the
+// node lists that are stored in PodScheduling objects. This limit is part
+// of the API.
+const PodSchedulingNodeListMaxSize = 128
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PodSchedulingList is a collection of pod scheduling objects.
+type PodSchedulingList struct {
+	metav1.TypeMeta
+	// Standard list metadata
+	metav1.ListMeta
+
+	// Items is the list of resource claims.
+	Items []PodScheduling
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ResourceClass is used by administrators to influence how resources
+// are allocated.
+//
+// This is an alpha type and requires enabling the DynamicResourceAllocation
+// feature gate.
+type ResourceClass struct {
+	metav1.TypeMeta
+	// Standard object metadata
+	metav1.ObjectMeta
+
+	// DriverName determines which resource driver is to be used for
+	// allocation of a ResourceClaim that uses this class.
+	//
+	// Resource drivers have a unique name in reverse domain order
+	// (acme.example.com).
+	DriverName string
+
+	// ParametersRef references an arbitrary separate object that may hold
+	// parameters that will be used by the driver when allocating a
+	// resource that uses this class. The driver will be able to
+	// distinguish between parameters stored here and and those stored in
+	// ResourceClaimSpec. These parameters here can only be set by cluster
+	// administrators.
+	ParametersRef *ResourceClassParametersReference
+
+	// Only nodes matching the selector will be considered by the scheduler
+	// when trying to find a Node that fits a Pod when that Pod uses
+	// a ResourceClaim that has not been allocated yet.
+	//
+	// Setting this field is optional. If null, all nodes are candidates.
```

Comment:
```suggestion
	// A list of node names. When a scheduler is filtering nodes for placing an incoming
	// Pod, and that Pod uses a dynamic resource from this ResourceClass, the scheduler
	// should treat nodes not on this list as excluded.
	//
	// If this list is empty, a scheduler considers all nodes as potential candidates.
```

Aside: why isn't this either:
- a list of _selectors_?
- part of the `status` of the resource class.

Reply:
The intended usage is to select nodes by labels where a tool like NFD sets the labels based on hardware characteristics of the node. The resource driver itself then gets deployed so that it only runs on those nodes.

The set of nodes can change, therefore a list of node names isn't suitable for this purpose. Some controller could update the information in status, but for dynamic selection we have PodScheduling.